## CNS_NoSpec

Compiler: gcc, gfortran  
gcc flags: -std=c99 -O2 -ftree-vectorize -DBL_Linux -DBL_FORT_USE_UNDERSCORE  
gfortran flags: -O2 -ftree-vectorize -fno-range-check  
On Intel(R) Xeon(R) CPU E5-2699 v4 @ 2.20GHz (Broadwell)  

#### CPU Time
The benchmark exclusive cputime breaks down largely to 3 files.  
  
60.2% CPU Time exclusively in advance.f90  
18.9% CPU Time exclusively in libc  
15.0% CPU Time exclusively in fabio.f90

#### advance.f90 CPU Time
29.2% Total CPU Time spent exclusively in `diffterm` subroutine.  8 Loops  
19.1% Total CPU Time spent exclusively in `hypterm` subroutine.  3 Loops  
11.8% Total CPU Time spent exclusively in `ctoprim` subroutine.  2 Loops  
  


Example Loop: `diffterm` subroutine in advance.f90  
5.6% Total CPU Time spent exclusively in this loop.  
```fortran
    !$OMP PARALLEL PRIVATE(i,j,k)
    !$OMP DO
    do k=lo(3)-ng,hi(3)+ng
       do j=lo(2)-ng,hi(2)+ng
          do i=lo(1),hi(1)

             ux(i,j,k)= &
                   (ALP*(q(i+1,j,k,qu)-q(i-1,j,k,qu)) &
                  + BET*(q(i+2,j,k,qu)-q(i-2,j,k,qu)) &
                  + GAM*(q(i+3,j,k,qu)-q(i-3,j,k,qu)) &
                  + DEL*(q(i+4,j,k,qu)-q(i-4,j,k,qu)))*dxinv(1)

             vx(i,j,k)= &
                   (ALP*(q(i+1,j,k,qv)-q(i-1,j,k,qv)) &
                  + BET*(q(i+2,j,k,qv)-q(i-2,j,k,qv)) &
                  + GAM*(q(i+3,j,k,qv)-q(i-3,j,k,qv)) &
                  + DEL*(q(i+4,j,k,qv)-q(i-4,j,k,qv)))*dxinv(1)

             wx(i,j,k)= &
                   (ALP*(q(i+1,j,k,qw)-q(i-1,j,k,qw)) &
                  + BET*(q(i+2,j,k,qw)-q(i-2,j,k,qw)) &
                  + GAM*(q(i+3,j,k,qw)-q(i-3,j,k,qw)) &
                  + DEL*(q(i+4,j,k,qw)-q(i-4,j,k,qw)))*dxinv(1)
          enddo
       enddo
    enddo
```

####  Subroutine Instruction Breakdown

| Subroutine | Load Inst. | Store Inst. | Branch Inst. |
|:----------:|:----------:|:-----------:|:------------:|
| diffterm   |  90.7%     | 8.3%        | 1.0%         |
| hypterm    |  87.8%     | 10.5%       | 1.7%         |
| ctoprim    |  70.8%     | 16.0%       | 13.2%        |

### L1 Data Cache
Hit Rate = 1 - (Misses /(Load Inst + Store Inst))

In [16]:
1 - (4.5 / (7.72 + 5.16))

0.6506211180124224

Aggregate L1 Data Cache Hit Rate: 65.1%   
  
`diffterm` subroutine L1 Data Cache Hit Rate: 69.3%  
`hypterm` subroutine L1 Data Cache Hit Rate: 65.2%  
`ctoprim` subroutine L1 Data Cache Hit Rate: 51.4%